message color
{
  required float  r			  =  1;
  required float  g			  =  2;
  required float  b			  =  3;

}

message Debug_Point
{
  required float  x           =  1;
  required float  y           =  2;
  optional color  col         =  3;

}

message Debug_Line
{
  required float  x1          =  1;
  required float  y1          =  2;
  required float  x2          =  3;
  required float  y2          =  4;
  optional color  col         =  5;
}

message Debug_Rect
{
  required float  x          =  1;
  required float  y          =  2;
  required float  w          =  3;
  required float  h          =  4;
  optional color  col        =  5;

}

message Debug_Circle
{
  required float  x          =  1;
  required float  y          =  2;
  required float  r          =  3;
  optional color  col        =  4;

}

enum Data_Type {
	imm_bool    = 0;
    imm_short   = 1;
	imm_int     = 2;
	imm_float   = 3;
	imm_double  = 4;
	imm_vec2    = 5;
	imm_vec3    = 6;
	imm_str     = 7;

  }

message Data_Node
{
  required string    name	 =  1;
  required Data_Type type	 =  2;
  required string    data	 =  3;
  optional string    hint    =  4;
}

message Debug_Draw
{
  repeated Debug_Point  point  = 1;
  repeated Debug_Line   line   = 2;
  repeated Debug_Rect   rect   = 3;
  repeated Debug_Circle circle = 4;

}

message Function_Data
{
  required string	  name         = 1;
  optional string	  category     = 2 [default = "Uncategorised"];
  optional string	  hint         = 3;
  repeated Data_Node  In_Data      = 4;
  repeated Data_Node  Out_Data     = 5;
  repeated Data_Node  Internal_Var = 6;
  required Debug_Draw draw_data    = 7;
}

message AI_Debug
{
  required uint32         frame_id = 1;
  repeated Function_Data  func     = 2;
  repeated Debug_Point  point  = 3;
  repeated Debug_Line   line   = 4;
  repeated Debug_Rect   rect   = 5;
  repeated Debug_Circle circle = 6;
}